unreliable HAVE_PTHREAD_TIMEDJOIN_NP check

build on FreeBSD 11 ends in exception:
`error: `pthread_timedjoin_np' was not declared in this scope`
The author of this [blog](http://e-mats.org/2012/01/building-gearman-under-red-hat-enterprise-4-rhel4-or-older-centos-versions/) described the same pain.

Is it an issue for [m4/ax_pthread_timedjoin_np.m4](https://github.com/gearman/gearmand/blob/master/m4/ax_pthread_timedjoin_np.m4)?

> It appears as though, on some FreeBSD versions, you need to link with -lthr to use pthread_timedjoin_np?

Unfortunately it doesn't work. I'm using FreeBSD 11.
```bash
CXX=c++49 CC=gcc LIBS="-lthr" ./configure --disable-option-checkin --with-postgresql=no --enable-shared=no
gmake
```
throws exception:
`libgearman-server/gearmand_thread.cc: In function 'void gearmand_thread_free(gearmand_thread_st*)':
libgearman-server/gearmand_thread.cc:295:68: error: 'pthread_timedjoin_np' was not declared in this scope
           pthread_error= pthread_timedjoin_np(thread->id, NULL, &ts);
`

The line [295](https://github.com/gearman/gearmand/blob/master/libgearman-server/gearmand_thread.cc#L295) lies in `#if defined(HAVE_PTHREAD_TIMEDJOIN_NP) && HAVE_PTHREAD_TIMEDJOIN_NP` block.

--- a/libgearman-server/gearmand_thread.cc
+++ b/libgearman-server/gearmand_thread.cc
@@ -50,6 +50,10 @@
 #include <memory>
 #include <csignal>
 
+#ifdef HAVE_PTHREAD_NP_H
+# include <pthread_np.h>
+#endif
+
 /*
  * Private declarations
  */

--- a/m4/ax_pthread_timedjoin_np.m4
+++ b/m4/ax_pthread_timedjoin_np.m4
@@ -19,11 +19,16 @@
 #   and this notice are preserved. This file is offered as-is, without any
 #   warranty.
 
-#serial 6
+#serial 7
 
 AC_DEFUN([AX_PTHREAD_TIMEDJOIN_NP],
     [AC_PREREQ([2.63])dnl
     AC_REQUIRE([AX_PTHREAD])
+
+    dnl FreeBSD keeps pthread_timedjoin_np's prototype in pthread_np.h,
+    dnl separate from pthread.h.
+    AC_CHECK_HEADERS([pthread_np.h])
+
     AC_CACHE_CHECK([check for pthread_timedjoin_np], [ax_cv_pthread_timedjoin_np],
       [AX_SAVE_FLAGS
       CFLAGS="$PTHREAD_CFLAGS"
@@ -32,13 +37,19 @@ AC_DEFUN([AX_PTHREAD_TIMEDJOIN_NP],
       AC_LINK_IFELSE(
         [AC_LANG_PROGRAM(
           [
+#ifndef _GNU_SOURCE
+# define _GNU_SOURCE
+#endif
 #include <pthread.h>
 #include <stdlib.h>
+#ifdef HAVE_PTHREAD_NP_H
+# include <pthread_np.h>
+#endif
           ],[
           pthread_t thread;
           pthread_timedjoin_np(thread, NULL, NULL);
           ])],
-        [ax_cv_pthread_timedjoin_np=yes],[])
+        [ax_cv_pthread_timedjoin_np=yes],[ax_cv_pthread_timedjoin_np=no])
 
       AC_LANG_POP
       AX_RESTORE_FLAGS

--- a/tests/hostile.cc
+++ b/tests/hostile.cc
@@ -60,6 +60,10 @@ using namespace org::gearmand;
 #include <unistd.h>
 #include <sys/time.h>
 
+#ifdef HAVE_PTHREAD_NP_H
+# include <pthread_np.h>
+#endif
+
 #define WORKER_FUNCTION_NAME "foo"
 
 #ifndef SERVER_TARGET

